src/backend/distributed/planner/multi_explain.c

```diff
@@ -639,6 +688,289 @@ BuildRemoteExplainQuery(char *queryString, ExplainState *es)
 }
 
 
+/*
+ * ShouldSaveWorkerQueryExplainAnalyze returns true if we should save the EXPLAIN
+ * ANALYZE output of given query while executing it.
+ */
+bool
+ShouldSaveWorkerQueryExplainAnalyze(QueryDesc *queryDesc)
+{
+	PlannedStmt *plannedStmt = queryDesc->plannedstmt;
+
+	return SaveTaskExplainPlans &&
+		   ExecutorLevel == 0 &&
+		   !IsParallelWorker() &&
+		   !IsCitusPlan(plannedStmt->planTree);
+}
+
+
+/*
+ * SaveQueryExplain saves the EXPLAIN ANALYZE output of an already executed
+ * and instrumented query. This can be retrieved later by calling
+ * worker_worker_last_saved_explain_analyze.
+ */
+void
+SaveQueryExplainAnalyze(QueryDesc *queryDesc)
+{
+	/*
+	 * Make sure stats accumulation is done.  (Note: it's okay if several
+	 * levels of hook all do this.)
+	 */
+	InstrEndLoop(queryDesc->totaltime);
+
+	ExplainState *es = NewExplainState();
+
+	es->analyze = true;
+	es->verbose = WorkerQueryExplainOptions.verbose;
+	es->buffers = WorkerQueryExplainOptions.buffers;
+	es->timing = WorkerQueryExplainOptions.timing;
+	es->summary = WorkerQueryExplainOptions.summary;
+	es->format = WorkerQueryExplainOptions.format;
+	es->costs = WorkerQueryExplainOptions.costs;
+
+	ExplainBeginOutput(es);
+
+	if (WorkerQueryExplainOptions.query)
+	{
+		ExplainQueryText(es, queryDesc);
+	}
+
+	ExplainPrintPlan(es, queryDesc);
+
+	if (es->costs)
+	{
+		ExplainPrintJITSummary(es, queryDesc);
+	}
+
+	ExplainEndOutput(es);
+
+	/*
+	 * We don't always send BEGIN/END to workers for all queries. Changing
+	 * transaction management for EXPLAIN ANALYZE can change their runtime
+	 * metrics, so we don't want to change it for more realistic EXPLAIN
+	 * ANALYZE.
```

Comment:
True, but the current approach breaks pretty badly when there are intermediate pgbouncers, so maybe we'll just have to accept using transaction blocks, which will also simplify memory management.